esp_flasher/gui.py

```
@@ -7,16 +7,389 @@
+    def _start_improv(self):
+        """Start Improv on the already-open serial port (same as JS: port stays open)."""
+        if not self._serial_port or not self._serial_port.is_open:
+            self.status_label.setText("Serial port not open")
+            return
+
+        # Drain & discard stale console data so the Improv state-machine
+        # starts clean.  The read-loop also handles stale bytes via its
+        # newline-reset logic, so this is belt-and-suspenders.
+        try:
+            self._serial_port.reset_input_buffer()
+        except (OSError, serial.SerialException) as e:
```

Ruff (0.15.9)

[error] Undefined name `serial`

(F821)